src/Generation/EmulatorSupportGenerator.php

```diff
@@ -0,0 +1,95 @@
+use Google\Generator\Utils\Type;
+
+class EmulatorSupportGenerator
+{
+    /**
+     * Map of emulator support required clients and their expected env variables.
+     */
+    private static $emulatorSupportClients = [
+        '\Google\Cloud\Spanner\Admin\Database\V1\Client\DatabaseAdminClient' => 'SPANNER_EMULATOR_HOST',
+        '\Google\Cloud\Spanner\Admin\Instance\V1\Client\InstanceAdminClient' => 'SPANNER_EMULATOR_HOST',
+        // Added for unittesting
+        '\Testing\Basic\Client\BasicClient' => 'BASIC_EMULATOR_HOST'
+    ];
+
+    /** @var string Name of the default emulator config function. */
+    public const DEFAULT_EMULATOR_CONFIG_FN = 'getDefaultEmulatorConfig';
+
+    public static function generateEmulatorSupportIfRequired(ServiceDetails $serviceDetails, SourceFileContext $ctx) {
+        $fullClassName = $serviceDetails->gapicClientV2Type->getFullName();
+        $emulatorHostVar = AST::var('emulatorHost');
+        $phpUrlSchemeConst = AST::constant('PHP_URL_SCHEME');
+        $schemeVar = AST::var('scheme');
+        $searchVar = AST::var('search');
+
+        if (array_key_exists($fullClassName, self::$emulatorSupportClients)) {
+            return AST::method(self::DEFAULT_EMULATOR_CONFIG_FN)
+            ->withAccess(Access::PRIVATE)
+            ->withBody(AST::block(
+                AST::assign($emulatorHostVar, AST::call(AST::GET_ENV)(self::$emulatorSupportClients[$fullClassName])),
+                AST::if(AST::call(AST::EMPTY)($emulatorHostVar))->then(AST::return(AST::array([]))),
+                AST::if(AST::assign($schemeVar, AST::call(AST::PARSE_URL)($emulatorHostVar, $phpUrlSchemeConst)))
+                    ->then(AST::block(
+                        AST::assign($searchVar, AST::binaryOp($schemeVar, '.', '://')),
+                        AST::assign($emulatorHostVar, AST::call(AST::STRING_REPLACE)($searchVar, '', $emulatorHostVar)),
+                    )),
+                AST::return(
+                    AST::array([
+                        'apiEndpoint' => $emulatorHostVar,
+                        'transportConfig' => [
+                            'grpc' => [
+                                'stubOpts' => [
+                                    'credentials' => AST::staticCall(
+                                        $ctx->type((Type::fromName("Grpc\ChannelCredentials"))),
+                                        AST::method('createInsecure'))()
+                                ]
+                            ]
+                        ],
+                        'credentials' => AST::new($ctx->type((Type::fromName("Google\ApiCore\InsecureCredentialsWrapper"))))(),
+                    ]))), AST::return(AST::array([])))
+            ->withPhpDoc(PhpDoc::block(
+                PhpDoc::text("Configure the gapic configuration to use a service emulator."),
+            ))
+            ->withReturnType($ctx->type(Type::array()));
+        } else {
+            return null;
+        }
```

Comment:
Invert this, return early if not configured. Exiting early when a condition isn't met is easier to read top-to-bottom, and dedents the code (easier on the eyes)
```suggestion
        if (!array_key_exists($fullClassName, self::$emulatorSupportClients)) {
            return null;
        }

        return AST::method(self::DEFAULT_EMULATOR_CONFIG_FN)
            ->withAccess(Access::PRIVATE)
            ->withBody(AST::block(
                AST::assign($emulatorHostVar, AST::call(AST::GET_ENV)(self::$emulatorSupportClients[$fullClassName])),
                AST::if(AST::call(AST::EMPTY)($emulatorHostVar))->then(AST::return(AST::array([]))),
                AST::if(AST::assign($schemeVar, AST::call(AST::PARSE_URL)($emulatorHostVar, $phpUrlSchemeConst)))
                    ->then(AST::block(
                        AST::assign($searchVar, AST::binaryOp($schemeVar, '.', '://')),
                        AST::assign($emulatorHostVar, AST::call(AST::STRING_REPLACE)($searchVar, '', $emulatorHostVar)),
                    )),
                AST::return(
                    AST::array([
                        'apiEndpoint' => $emulatorHostVar,
                        'transportConfig' => [
                            'grpc' => [
                                'stubOpts' => [
                                    'credentials' => AST::staticCall(
                                        $ctx->type((Type::fromName("Grpc\ChannelCredentials"))),
                                        AST::method('createInsecure'))()
                                ]
                            ]
                        ],
                        'credentials' => AST::new($ctx->type((Type::fromName("Google\ApiCore\InsecureCredentialsWrapper"))))(),
                    ]))), AST::return(AST::array([])))
            ->withPhpDoc(PhpDoc::block(
                PhpDoc::text("Configure the gapic configuration to use a service emulator."),
            ))
            ->withReturnType($ctx->type(Type::array()));
```

You will need to fix tabbing from suggestion, or just edit yourself.

Reply:
Done